flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/BulkSlotProviderImpl.java

```diff
@@ -0,0 +1,204 @@
+
+		final Optional<PhysicalSlot> availablePhysicalSlot = tryAllocateFromAvailable(slotRequestId, slotProfile);
+
+		final CompletableFuture<PhysicalSlot> slotFuture;
+		if (availablePhysicalSlot.isPresent()) {
+			slotFuture = CompletableFuture.completedFuture(availablePhysicalSlot.get());
+		} else if (physicalSlotRequest.willSlotBeOccupiedIndefinitely()) {
+			slotFuture = slotPool.requestNewAllocatedSlot(
+				slotRequestId,
+				resourceProfile,
+				null);
+		} else {
+			slotFuture = slotPool.requestNewAllocatedBatchSlot(
+				slotRequestId,
+				resourceProfile);
+		}
+
+		return slotFuture.thenApply(physicalSlot -> new PhysicalSlotRequest.Result(slotRequestId, physicalSlot));
+	}
+
+	private Optional<PhysicalSlot> tryAllocateFromAvailable(
+			final SlotRequestId slotRequestId,
+			final SlotProfile slotProfile) {
+
+		final Collection<SlotSelectionStrategy.SlotInfoAndResources> slotInfoList =
+			slotPool.getAvailableSlotsInformation()
+				.stream()
+				.map(SlotSelectionStrategy.SlotInfoAndResources::fromSingleSlot)
+				.collect(Collectors.toList());
+
+		final Optional<SlotSelectionStrategy.SlotInfoAndLocality> selectedAvailableSlot =
+			slotSelectionStrategy.selectBestSlotForProfile(slotInfoList, slotProfile);
+
+		return selectedAvailableSlot.flatMap(
+			slotInfoAndLocality -> slotPool.allocateAvailableSlot(
+				slotRequestId,
+				slotInfoAndLocality.getSlotInfo().getAllocationId())
+		);
+	}
+
+	private void schedulePendingRequestBulkTimeoutCheck(
+			final PhysicalSlotRequestBulk slotRequestBulk,
+			final Time timeout) {
+
+		componentMainThreadExecutor.schedule(() -> {
+			final PhysicalSlotRequestBulkTracker.TimeoutCheckResult result =
+				slotRequestBulkTracker.checkPhysicalSlotRequestBulkTimeout(slotRequestBulk, timeout);
+
+			switch (result) {
+				case PENDING:
+					//re-schedule the timeout check
+					schedulePendingRequestBulkTimeoutCheck(slotRequestBulk, timeout);
+					break;
+				case TIMEOUT:
+					timeoutSlotRequestBulk(slotRequestBulk);
+					break;
+				default: // no action to take
+			}
+		}, timeout.getSize(), timeout.getUnit());
+	}
+
+	private void timeoutSlotRequestBulk(final PhysicalSlotRequestBulk slotRequestBulk) {
+		final Exception cause = new TimeoutException("Slot request bulk is not fulfillable!");
+		// pending requests must be canceled first otherwise they might be fulfilled by
+		// allocated slots released from this bulk
+		for (SlotRequestId slotRequestId : slotRequestBulk.getPendingRequests().keySet()) {
+			cancelSlotRequest(slotRequestId, cause);
+		}
+		for (SlotRequestId slotRequestId : slotRequestBulk.getFulfilledRequests().keySet()) {
+			cancelSlotRequest(slotRequestId, cause);
+		}
+	}
+
+	private Set<SlotInfo> getAllSlotInfos() {
+		return Stream
+			.concat(
+				slotPool.getAvailableSlotsInformation().stream(),
+				slotPool.getAllocatedSlotsInformation().stream())
+			.collect(Collectors.toSet());
+	}
```

Comment:
I thought we could actually move `timeoutSlotRequestBulk/getAllSlotInfos` and pass `slotPool` to `PhysicalSlotRequestBulkTracker`. Would this simplify multithreading in tests?
For multithreading, we would only need to check whether `BulkSlotProviderImpl` properly schedules `checkPhysicalSlotRequestBulkTimeout`.

Reply:
Sorry I do not quite understand what do you mean by 'simplify multithreading'?

Currently we are a `slotsRetriever` to `PhysicalSlotRequestBulkTracker`. It simplifies testing and hides actions methods of `SlotPool` that a `PhysicalSlotRequestBulkTracker` should not know.
`SchedulerImpl` invokes `checkPhysicalSlotRequestBulkTimeout` and I think it's better to let it decide how to deal with the check results. That's why I put `timeoutSlotRequestBulk` in `SchedulerImpl`.

Reply:
`PhysicalSlotRequestBulkTracker` does not contains actions on slot requests. All actions (allocating/canceling) take place in `BulkSlotProviderImpl`. I think this makes it easier to reasoning the slot status.

Reply:
I thought to separate slot request submission (`BulkSlotProviderImpl`) and tracking (`PhysicalSlotRequestBulkTracker`). On the other hand, it is easier to test the current `PhysicalSlotRequestBulkTracker` w/o `SlotPool`. Thanks for explanation.